sdk-extension/opentelemetry-sdk-extension-aws/src/opentelemetry/sdk/extension/aws/trace/propagation/aws_xray_format.py

```diff
@@ -0,0 +1,283 @@
+
+import logging
+import typing
+
+import opentelemetry.trace as trace
+from opentelemetry.context import Context
+from opentelemetry.trace.propagation.textmap import (
+    Getter,
+    Setter,
+    TextMapPropagator,
+    TextMapPropagatorT,
+)
+
+_logger = logging.getLogger(__name__)
+
+
+class AwsXRayFormat(TextMapPropagator):
+    """Propagator for the AWS X-Ray Trace Header propagation protocol.
+
+    See: https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader
+    """
+
+    # AWS
+    TRACE_HEADER_KEY = "X-Amzn-Trace-Id"
+
+    KV_PAIR_DELIMITER = ";"
+    KEY_AND_VALUE_DELIMITER = "="
+
+    TRACE_ID_KEY = "Root"
+    TRACE_ID_LENGTH = 35
+    TRACE_ID_VERSION = "1"
+    TRACE_ID_DELIMITER = "-"
+    TRACE_ID_DELIMITER_INDEX_1 = 1
+    TRACE_ID_DELIMITER_INDEX_2 = 10
+    TRACE_ID_FIRST_PART_LENGTH = 8
+
+    PARENT_ID_KEY = "Parent"
+    PARENT_ID_LENGTH = 16
+
+    SAMPLED_FLAG_KEY = "Sampled"
+    SAMPLED_FLAG_LENGTH = 1
+    IS_SAMPLED = "1"
+    NOT_SAMPLED = "0"
+
+    def extract(
+        self,
+        getter: Getter[TextMapPropagatorT],
+        carrier: TextMapPropagatorT,
+        context: typing.Optional[Context] = None,
+    ) -> Context:
+        trace_header_list = getter(carrier, self.TRACE_HEADER_KEY)
+
+        if not trace_header_list or len(trace_header_list) != 1:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        trace_header = trace_header_list[0]
+
+        if not trace_header:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        trace_id, span_id, sampled, err = self.extract_span_properties(
+            trace_header
+        )
+
+        if err is not None:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        options = 0
+        if sampled:
+            options |= trace.TraceFlags.SAMPLED
+
+        span_context = trace.SpanContext(
+            trace_id=trace_id,
+            span_id=span_id,
+            is_remote=True,
+            trace_flags=trace.TraceFlags(options),
+            trace_state=trace.TraceState(),
+        )
+
+        if not span_context.is_valid:
+            _logger.error(
+                "Invalid Span Extracted. Insertting INVALID span into provided context."
+            )
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        return trace.set_span_in_context(
+            trace.DefaultSpan(span_context), context=context
+        )
+
+    def extract_span_properties(self, trace_header):
+        trace_id = trace.INVALID_TRACE_ID
+        span_id = trace.INVALID_SPAN_ID
+        sampled = False
+
+        extract_err = None
+
+        for kv_pair_str in trace_header.split(self.KV_PAIR_DELIMITER):
+            if extract_err:
```

Comment:
it looks like this extract_err is defensive, since there is an appropriate break call in all the other branches.

I may consider not including this one. It was a little confusing reading through this to understand the value of this block since it's handled correctly elsewhere.

Reply:
nevermind, I see that there's a couple branches that require this block.